conda/cli/install.py

```diff
@@ -309,12 +316,75 @@ def ensure_update_specs_exist(prefix: str, specs: list[str]):
         spec = MatchSpec(spec)
         if not spec.is_name_only_spec:
             raise CondaError(
-                f"Invalid spec for 'conda update': {spec}\nUse 'conda install' instead."
+                f"'conda update' only supports name-only spec, but received: {spec}\n"
+                f"Use 'conda install' to install a specific version of a package."
             )
         if not prefix_data.get(spec.name, None):
             raise PackageNotInstalledError(prefix, spec.name)
 
 
+def _assemble_environment(
+    name: str | None = None,
+    prefix: str | None = None,
+    specs: list[str] = (),
+    files: list[str] = (),
+    inject_default_packages: bool = True,
+) -> Environment:
+    # First, let's create an 'Environment' for the information exposed in the CLI (no files)
+    if inject_default_packages:
+        names = {MatchSpec(pkg).name for pkg in specs}
+        for pkg in context.create_default_packages:
+            pkg_name = MatchSpec(pkg).name
+            if pkg_name not in names:
+                specs.append(pkg)
+
+    requested_packages = []
+    fetch_explicit_packages = []
+
+    # extract specs from files
+    for fpath in files:
+        try:
+            specs.extend(
+                [
+                    spec
+                    for spec in common.specs_from_url(fpath)
+                    if spec != EXPLICIT_MARKER
+                ]
+            )
+        except UnicodeError:
+            raise CondaError(
+                "Error reading file, file should be a text file containing"
+                " packages \nconda create --help for details"
+            )
+
+    for spec in specs:
```

Comment:
If we filter out default packages based on `specs` shouldn't we do the same for specs defined in `files`?

Reply:
Good catch! This is consistent with the behaviour on the latest release of conda  (and older versions, I tested as old as 24.7.0). tbh I'm not sure if this is a bug or not, but I opened up a bug report in https://github.com/conda/conda-environments-project-mgmt/issues/86. Would you call this a bug?

Reply:
Cool! Since this is a bug that already exists in the currently released conda, I would suggest fixing it in a follow up PR. That'll keep this PR lean and focused and (hopefully) easier to review.